linkerd/retry/src/replay.rs

```diff
@@ -0,0 +1,766 @@
+            "Replay::poll_data"
+        );
+
+        // If we haven't replayed the buffer yet, and its not empty, return the
+        // buffered data first.
+        if this.replay_body && state.buf.has_remaining() {
+            tracing::trace!("replaying body");
+            // Don't return the buffered data again on the next poll.
+            this.replay_body = false;
+            return Poll::Ready(Some(Ok(Data::Replay(state.buf.clone()))));
+        }
+
+        // If the inner body has previously ended, don't poll it again.
+        //
+        // NOTE(eliza): we would expect the inner body to just happily return
+        // `None` multiple times here, but `hyper::Body::channel` (which we use
+        // in the tests) will panic if it is polled after returning `None`, so
+        // we have to special-case this. :/
+        if state.is_completed {
+            return Poll::Ready(None);
+        }
+
+        // If there's more data in the initial body, poll that...
+        if let Some(rest) = state.rest.as_mut() {
+            tracing::trace!("Polling initial body");
+            let opt = futures::ready!(Pin::new(rest).poll_data(cx));
+
+            // If the body
+            if opt.is_none() {
+                tracing::trace!("Initial body completed");
+                state.is_completed = true;
+            }
+            return Poll::Ready(
+                opt.map(|ok| ok.map(|data| Data::Initial(state.buf.push_chunk(data)))),
+            );
+        }
+
+        // Otherwise, guess we're done!
+        Poll::Ready(None)
+    }
+
+    fn poll_trailers(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Result<Option<HeaderMap>, Self::Error>> {
+        let this = self.get_mut();
+        let state = Self::state(&mut this.state, &this.shared);
+        tracing::trace!(
+            replay_trailers = this.replay_trailers,
+            "Replay::poll_trailers"
+        );
+
+        if this.replay_trailers {
+            this.replay_trailers = false;
+            if let Some(ref trailers) = state.trailers {
+                tracing::trace!("Replaying trailers");
+                return Poll::Ready(Ok(Some(trailers.clone())));
+            }
+        }
+
+        if let Some(rest) = state.rest.as_mut() {
+            // If the inner body has previously ended, don't poll it again.
+            if !rest.is_end_stream() {
+                let res = futures::ready!(Pin::new(rest).poll_trailers(cx)).map(|tlrs| {
+                    if state.trailers.is_none() {
+                        state.trailers = tlrs.clone();
+                    }
+                    tlrs
+                });
+                return Poll::Ready(res);
+            }
+        }
+
+        Poll::Ready(Ok(None))
+    }
+
+    fn is_end_stream(&self) -> bool {
+        self.state
+            .as_ref()
+            .and_then(|state| state.rest.as_ref().map(Body::is_end_stream))
+            .unwrap_or(false)
+    }
```

Comment:
If I understand correctly this will always return `false` unless `poll_data` is called first, and so checks like that in clone_request may not work. What if the underlying body is empty? I suppose it's not a _bug_ for this return false by default; but it also seems like we could preserve the initial value... not sure.

In any case, maybe worth a comment explaining the thought process and/or test.

Reply:
Oh! It seems like we always start with a `state` when we construct the first body, and only lazily acquire it in clones;  It's probably fine, then.

Reply:
> Oh! It seems like we always start with a `state` when we construct the first body, and only lazily acquire it in clones; It's probably fine, then.

Yeah, that's right...I can add comments explaining this a little better?

Reply:
Ah, I wrote a test for this and it is, actually, wrong --- since shared state isn't _acquired_ here, we do return `false` until the clone is polled for the first time. I'll fix that.

Reply:
...although, actually, this logic is _quite_ wrong, since it will also return `true` if the actual body has been polled to completion but a given clone hasn't been polled yet. oops!